File: src/main/java/saru/view/LadderOutput.java
package saru.view;

import saru.domain.LadderGame;
import saru.domain.Line;
import saru.domain.User;

import java.util.List;
import java.util.Scanner;

public class LadderOutput {
    private static final Scanner scanner = new Scanner(System.in);

    private static final String RESULT_PROMPT = "결과를 보고 싶은 사람은?";
    private static final String ROW_BLANK_LINE = "     ";
    private static final String COLUMN_LINE = "|";
    private static final String ROW_LINE = "-----";

    public LadderOutput() {
    }

    public void sendOutputObject(LadderGame ladderGame) {
        printUserNames(ladderGame.getUsers());
        printLadder(ladderGame.getLadderLines());
        printDestination(ladderGame.getDestinations());

        boolean isAllResult = false;
        while (!isAllResult) {
            isAllResult = promptResult(ladderGame);
        }
    }

    private void printUserNames(List<User> users) {
        for (User user : users) {
            System.out.printf("%-6s", user.getName());
        }
        System.out.println();
    }

    // TODO 리팩토링 필요
    private boolean promptResult(LadderGame ladderGame) {
        System.out.println(RESULT_PROMPT);
        String userChoice = getUserChoice();

        List<User> users = ladderGame.getUsers();
        List<Integer> results = ladderGame.getResult();
        List<String> destination = ladderGame.getDestinations();

        if (userChoice.equals("all")) {
            printAllResult(users, results, destination);
            return true;
        }

        int matchIndex = getUserMatchIndex(users, userChoice);
        System.out.println(destination.get(results.get(matchIndex) / 2));

        return false;
    }

    // TODO 인덴트를 못 줄이겠습니다.
    private int getUserMatchIndex(List<User> users, String userName) {

        for (int i = 0; i < users.size(); i++) {
            if (users.get(i).getName().equals(userName)) {
                return i;
            }
        }

        return -1;
    }

    private void printAllResult(List<User> users, List<Integer> results, List<String> destination) {
        for (int i = 0; i < results.size(); i++) {
            System.out.println(users.get(i).getName() + " : " + destination.get(results.get(i) / 2));
        }
    }

    private String getUserChoice() {
        return scanner.nextLine();
    }

    private void printDestination(List<String> destinations) {
        for (String destination : destinations) {
            System.out.printf("%-6s", destination);
        }
        System.out.print("\n\n");
    }

    private void printLadder(List<Line> lines) {
        for (Line line : lines) {
            printMultiLines(line.getPoints());
        }
    }

    private void printMultiLines(List<Boolean> arr) {
        for (int i = 0; i < arr.size(); i++) {
            printIndividualLine(arr, i);
        }
        System.out.println();
    }

    private void printIndividualLine(List<Boolean> arr, int index) {
        if (index % 2 == 1) {
            printInterPoint(arr.get(index));
            return;
        }
        System.out.print(COLUMN_LINE);
    }

    private void printInterPoint(boolean isTrue) {
        if (isTrue) {
            System.out.print(ROW_LINE);
            return;
        }
        System.out.print(ROW_BLANK_LINE);
    }
}

File: src/main/kotlin/saru/view/LadderOutput.kt
package saru.view

import saru.domain.LadderGame
import saru.domain.Line
import saru.domain.User
import java.util.*

class LadderOutput {

    private val userChoice: String
        get() = scanner.nextLine()

    fun sendOutputObject(ladderGame: LadderGame) {
        printUserNames(ladderGame.getUsers())
        printLadder(ladderGame.getLadderLines())
        printDestination(ladderGame.destinations)

        var isAllResult = false
        while (!isAllResult) {
            isAllResult = promptResult(ladderGame)
        }
    }

    private fun printUserNames(users: List<User>) {
        for (user in users) {
            System.out.printf("%-6s", user.name)
        }
        println()
    }

    // TODO 리팩토링 필요
    private fun promptResult(ladderGame: LadderGame): Boolean {
        println(RESULT_PROMPT)
        val userChoice = userChoice

        val users = ladderGame.getUsers()
        val results = ladderGame.result
        val destination = ladderGame.destinations

        if (userChoice == "all") {
            printAllResult(users, results!!, destination)
            return true
        }

        val matchIndex = getUserMatchIndex(users, userChoice)
        println(destination[results!![matchIndex] / 2])

        return false
    }

    // TODO 인덴트를 못 줄이겠습니다.
    private fun getUserMatchIndex(users: List<User>, userName: String): Int {

        for (i in users.indices) {
            if (users[i].name == userName) {
                return i
            }
        }

        return -1
    }

    private fun printAllResult(users: List<User>, results: List<Int>, destination: List<String>) {
        for (i in results.indices) {
            println(users[i].name + " : " + destination[results[i] / 2])
        }
    }

    private fun printDestination(destinations: List<String>) {
        for (destination in destinations) {
            System.out.printf("%-6s", destination)
        }
        print("\n\n")
    }

    private fun printLadder(lines: List<Line>) {
        for (line in lines) {
            printMultiLines(line.getPoints())
        }
    }

    private fun printMultiLines(arr: List<Boolean>) {
        for (i in arr.indices) {
            printIndividualLine(arr, i)
        }
        println()
    }

    private fun printIndividualLine(arr: List<Boolean>, index: Int) {
        if (index % 2 == 1) {
            printInterPoint(arr[index])
            return
        }
        print(COLUMN_LINE)
    }

    private fun printInterPoint(isTrue: Boolean) {
        if (isTrue) {
            print(ROW_LINE)
            return
        }
        print(ROW_BLANK_LINE)
    }

    companion object {
        private val scanner = Scanner(System.`in`)

        private val RESULT_PROMPT = "결과를 보고 싶은 사람은?"
        private val ROW_BLANK_LINE = "     "
        private val COLUMN_LINE = "|"
        private val ROW_LINE = "-----"
    }
}
